File: Rotating the Box/Rotating the Box.py
# Solution for Rotating the Box in PY

class Solution:
    def rotateTheBox(self, boxGrid: List[List[str]]) -> List[List[str]]:
        
        originalRows = len(boxGrid)
        originalCols = len(boxGrid[0])

        # Create rotated matrix
        rotatedBox = [[None] * originalRows for _ in range(originalCols)]

        # STEP 1: Rotate the matrix
        for row in range(originalRows):
            for col in range(originalCols):
                rotatedBox[col][originalRows - row - 1] = boxGrid[row][col]

        # STEP 2: Apply gravity
        for col in range(originalRows):

            emptyRowPointer = originalCols - 1

            for row in range(originalCols - 1, -1, -1):

                if rotatedBox[row][col] == '*':
                    emptyRowPointer = row - 1
                elif rotatedBox[row][col] == '#':
                    rotatedBox[row][col] = '.'
                    rotatedBox[emptyRowPointer][col] = '#'
                    emptyRowPointer -= 1

        return rotatedBox

File: Rotating the Box/Rotating the Box.java
// Solution for Rotating the Box in JAVA

class Solution {
    public char[][] rotateTheBox(char[][] boxGrid) {

        int originalRows = boxGrid.length;
        int originalCols = boxGrid[0].length;

        // Create rotated matrix
        char[][] rotatedBox = new char[originalCols][originalRows];

        // STEP 1: Rotate the matrix
        for (int row = 0; row < originalRows; row++) {
            for (int col = 0; col < originalCols; col++) {
                rotatedBox[col][originalRows - row - 1] = boxGrid[row][col];
            }
        }

        // STEP 2: Apply gravity
        for (int col = 0; col < originalRows; col++) {

            int emptyRowPointer = originalCols - 1;

            for (int row = originalCols - 1; row >= 0; row--) {

                if (rotatedBox[row][col] == '*') {
                    emptyRowPointer = row - 1;
                } 
                else if (rotatedBox[row][col] == '#') {
                    rotatedBox[row][col] = '.';
                    rotatedBox[emptyRowPointer][col] = '#';
                    emptyRowPointer--;
                }
            }
        }

        return rotatedBox;
    }
}
